Use JSON format from asm_differ

zbanks opened the pull request:
This PR what I'd call the bare minimum to switch from the `asm_differ`'s `HtmlFormatter` to the `JsonFormatter`. 
I'm not very familiar with JSX -- so please critique the style/conventions there.

I'll leave some comments inline on the sections that are particularly *weird*.

![2021-09-16-155834_908x216_scrot](https://user-images.githubusercontent.com/113075/133677841-8ab15bad-06aa-4b69-9e5b-75820bff727f.png)

zbanks commented on frontend/src/scratch/Scratch.module.css:
```diff
@@ -24,35 +24,37 @@
 
 .diff {
     display: block;
+    border-spacing: 1em 0;
```
I wanted to add horizontal `border-spacing` to put some padding around the line number column. Setting the vertical spacing to `0` is actually less than we had before -- making the diff a bit denser. 

I like the denser diffs better, but I can imagine that the original line spacing was intentional? I'm fine reverting this, and putting the padding in `.diffLineNumber` instead.

nanaian replied:
> I can imagine that the original line spacing was intentional?

It wasn't! I'm happy with a denser diff personally.

zbanks commented on frontend/src/scratch/Scratch.module.css:
```diff
@@ -24,35 +24,37 @@
-.diff :global(.rotation-6) { color: lightcyan }
-.diff :global(.rotation-7) { color: lightgreen }
-.diff :global(.rotation-8) { color: grey }
+.diff-immediate { color: #6D6DFF }
+.diff-stack { color: #e3fc45 }
+.diff-register { color: #e3fc45 }
+.diff-delay_slot { font-weight: bold; color: #969896 }
+.diff-diff_change { color: #6D6DFF }
+.diff-diff_add { color: #45bd00 }
+.diff-diff_remove { color: #c82829 }
+.diff-source_filename { font-weight: bold }
+.diff-source_function { font-weight: bold; text-decoration: underline }
+.diff-source_other { font-style: italic }
+.diff-rotation-0 { color: magenta}
+.diff-rotation-1 { color: cyan }
+.diff-rotation-2 { color: rgb(0, 212, 0) }
+.diff-rotation-3 { color: red }
+.diff-rotation-4 { color: lightyellow }
+.diff-rotation-5 { color: lightpink }
+.diff-rotation-6 { color: lightcyan }
+.diff-rotation-7 { color: lightgreen }
+.diff-rotation-8 { color: grey }
+.diffLineNumber { color: grey; font-style: italic; font-size: 0.7em }
```
Okay, this is the weirdest mishmash of snake, kebab, and camel case that I've ever written :sweat_smile:. Not totally happy with it, but this is how I ended up here:

- I wanted to remove `:global`, now that we can use the standard `{style.foo}`
- I wanted to keep the formats used for the diff output namespaced, in case `asm_differ` e.g. if the `asm_differ` added a new `log` format in the future. This is why I prefixed all the classes used by asm_differ formats with `diff-`.
- The `asm_differ` formats use underscores (like python identifiers) instead of dashes. If we really hate the class names with mixed underscores/dashes, I could do `t.format.replace(/_/g, "-")`
- `.diffLineNumber` is camelCase because it's not a format from `asm_differ`, so it follows the default convention in this file.

nanaian replied:
I'd suggest writing a `Diff.module.css` and use camelCase instead in-keeping with other files.
I'm pretty sure I implemented this in #122 but that's sadly not merged yet

nanaian commented on backend/coreapp/asm_diff_wrapper.py:
```diff
@@ -103,16 +105,20 @@ def diff(target_assembly: Assembly, compilation: Compilation) -> str:
         display = Display(basedump, mydump, config)
 
-        return display.run_diff()[0]
+        # TODO: It would be nice to get a python object from `run_diff()` to avoid the JSON roundtrip
```
Worth making an issue on asm-differ for this?

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -298,7 +310,25 @@ export default function Scratch() {
+                                        <table class={styles.diff}>
+                                            <tr>
+                                                <th>{formatDiffText(diff.header.base)}</th>
+                                                <th>{/* Line */}</th>
```
Why not show the line no here?

zbanks replied:
This is for the header row (`TARGET`, `CURRENT (100)`) -- since the line column is narrow, I didn't want to add a label at the top? (Unless you think I should? None of the other "sub columns" are labelled, like the offset)

nanaian replied:
Ah, I didn't realise this was the header row!

Although, thinking about it, it might be nice to explain what each "sub column" is, if there's space.

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -184,6 +184,18 @@ export default function Scratch() {
         debouncedCompile()
     }, [debouncedCompile, compiler?.compiler, compiler?.cc_opts])
 
+    const formatDiffText = (texts) => {
```
```suggestion
    const formatDiffText = texts => {
```

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -184,6 +184,18 @@ export default function Scratch() {
+                return <span>{t.text}</span>
+            }
+        })
+    }
```
This function can be raised into module-level scope instead of placing it inside `Scratch` IMO

nanaian reviewed (commented):
The code generally looks good to me! Could you write TypeScript types for `diff`?

However, I'm sometimes getting an error on the frontend when I try this:
```
Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'row.base.text')
```

I have no idea what exactly the cause is, but I can reproduce it with this C code:
```c
extern int g;

void func() {
    // parent scratch
    int a;
    if (a == 9) {
        g += 2;
    }
}
```

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -76,7 +76,7 @@ export default function Scratch() {
         try {
             setCurrentRequest("compile")
-            const { diff_output, errors } = await api.post(`/scratch/${slug}/compile`, {
+            const { diff_output, errors } : api.DiffResponse = await api.post(`/scratch/${slug}/compile`, {
```
```suggestion
            const { diff_output, errors }: api.DiffResponse = await api.post(`/scratch/${slug}/compile`, {
```

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -315,15 +303,15 @@ export default function Scratch() {
-                                                    <td>{formatDiffText(row.base.text)}</td>
-                                                    <td><span class={styles.diffLineNumber}>{row.current.src_line}</span></td>
-                                                    <td>{formatDiffText(row.current.text)}</td>
+                                                    <td>{(row.base) && <FormatDiffText texts={row.base.text} />}</td>
```
```suggestion
                                                    <td>{row.base && <FormatDiffText texts={row.base.text} />}</td>
```

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -315,15 +303,15 @@ export default function Scratch() {
-                                                    <td>{formatDiffText(row.current.text)}</td>
+                                                    <td>{(row.base) && <FormatDiffText texts={row.base.text} />}</td>
+                                                    <td><span class={styles.diffLineNumber}>{(row.current) && row.current.src_line}</span></td>
+                                                    <td>{(row.current) && <FormatDiffText texts={row.current.text} />}</td>
```
```suggestion
                                                    <td>{row.current && <FormatDiffText texts={row.current.text} />}</td>
```

nanaian commented on frontend/src/scratch/Scratch.tsx:
```diff
@@ -315,15 +303,15 @@ export default function Scratch() {
-                                                    <td><span class={styles.diffLineNumber}>{row.current.src_line}</span></td>
-                                                    <td>{formatDiffText(row.current.text)}</td>
+                                                    <td>{(row.base) && <FormatDiffText texts={row.base.text} />}</td>
+                                                    <td><span class={styles.diffLineNumber}>{(row.current) && row.current.src_line}</span></td>
```
```suggestion
                                                    <td><span class={styles.diffLineNumber}>{row.current && row.current.src_line}</span></td>
```

nanaian reviewed (changes requested):
LGTM besides a few style nits

zbanks commented:
- Fixed the style nits (and now `yarn lint` covers more formatting stuff)
- Merged with Alex's PR from yesterday, #122, so now the `Diff` component is better abstracted/isolated.
- Updated the `asm_differ` subrepo to pull in the recent PR, which let me clean up `asm_diff_wrapper.py` slightly.
    - Now supports the `.data` references (for jump tables)

nanaian commented:
> and now yarn lint covers more formatting stuff

This was my change on main, correct? I don't see an eslint diff on this PR.

zbanks commented:
> > and now yarn lint covers more formatting stuff
> 
> This was my change on main, correct? I don't see an eslint diff on this PR.

Oh yeah - that was ambiguous. I meant to say: because of your change to main, there are now more automated warnings about formatting, therefore my code should have fewer formatting mistakes.

ethteck reviewed (approved):
I tested this locally, and it looks fantastic!

ethteck commented:
As per [discussion in discord ](https://discord.com/channels/688807550715560050/859699601971609610/889228057976668181), I'm gonna go ahead and merge this and we can continue design discussion in #115 .
